File: shared/src/main/java/com/guardpulse/parentcontrol/shared/PolicyConstants.kt
package com.guardpulse.parentcontrol.shared

object PolicyConstants {
    data class SettingsSectionPolicy(
        val packageName: String,
        val label: String,
        val shortLabel: String,
        val key: String
    )

    const val TV_PACKAGE = "com.guardpulse.parentcontrol.tv"
    const val PARENT_PACKAGE = "com.guardpulse.parentcontrol.parent"
    const val SETTINGS_APPS_PACKAGE = "com.guardpulse.policy.settings_apps"
    const val SETTINGS_DEVICE_PREFERENCES_PACKAGE = "com.guardpulse.policy.settings_device_preferences"
    const val SETTINGS_DEVELOPER_OPTIONS_PACKAGE = "com.guardpulse.policy.settings_developer_options"
    const val SETTINGS_SECURITY_RESTRICTIONS_PACKAGE = "com.guardpulse.policy.settings_security_restrictions"
    const val SETTINGS_ACCESSIBILITY_PACKAGE = "com.guardpulse.policy.settings_accessibility"
    const val SETTINGS_RESET_PACKAGE = "com.guardpulse.policy.settings_reset"
    const val DEPRECATED_SETTINGS_SECTIONS_PACKAGE = "com.guardpulse.policy.settings_sections"

    const val COMMAND_RESCAN_APPS = "rescanApps"
    const val COMMAND_RESET_TODAY = "resetToday"
    const val COMMAND_UNPAIR = "unpair"
    const val COMMAND_OPEN_SETUP = "openSetup"

    const val BLOCK_REASON_MANUAL = "manual"
    const val BLOCK_REASON_DAILY_LIMIT = "dailyLimit"
    const val BLOCK_REASON_RISKY_SETTINGS = "riskySettings"
    const val BLOCK_REASON_NETWORK_FILTER_MISSING = "networkFilterMissing"
    const val BLOCK_REASON_SOURCE_LOCK = "sourceLock"
    const val BLOCK_REASON_SETTINGS_SECTION = "settingsSection"

    const val ENFORCEMENT_DEVICE_OWNER = "deviceOwner"
    const val ENFORCEMENT_FALLBACK = "fallback"
    const val ENFORCEMENT_UNPROTECTED = "unprotected"

    const val UNLOCK_PENDING = "pending"
    const val UNLOCK_APPROVED = "approved"
    const val UNLOCK_DENIED = "denied"
    const val UNLOCK_EXPIRED = "expired"
    const val UNLOCK_APPROVAL_ONE_VISIT = "oneVisit"
    const val UNLOCK_APPROVAL_TIMED = "timed"

    const val SYNC_PROTOCOL_VERSION = 2
    const val SYNC_STATUS_APPLIED = "applied"
    const val SYNC_STATUS_FAILED = "failed"
    const val COMMAND_PENDING = "pending"
    const val COMMAND_RUNNING = "running"
    const val COMMAND_DONE = "done"
    const val COMMAND_FAILED = "failed"
    const val COMMAND_EXPIRED = "expired"

    const val PAIR_PENDING = "pending"
    const val PAIR_ACCEPTED = "accepted"
    const val PAIR_REJECTED = "rejected"
    const val PAIR_EXPIRED = "expired"
    const val PAIR_FAILED = "failed"

    const val REVISION_APP_POLICY = "appPolicy"
    const val REVISION_MODE_CREATE = "modeCreate"
    const val REVISION_MODE_UPDATE = "modeUpdate"
    const val REVISION_MODE_DELETE = "modeDelete"
    const val REVISION_MODE_POLICY = "modePolicy"
    const val REVISION_ACTIVE_MODE = "activeMode"
    const val REVISION_SAFE_MODE = "safeMode"
    const val REVISION_PIN = "pin"
    const val REVISION_MIGRATION = "migration"

    const val TAMPER_ADMIN_DISABLE_REQUESTED = "adminDisableRequested"
    const val TAMPER_ADMIN_DISABLED = "adminDisabled"
    const val TAMPER_ACCESSIBILITY_DISABLED = "accessibilityDisabled"
    const val TAMPER_USAGE_ACCESS_MISSING = "usageAccessMissing"
    const val TAMPER_VPN_DISABLED = "vpnDisabled"
    const val TAMPER_RISKY_SETTINGS_OPENED = "riskySettingsOpened"
    const val TAMPER_PIN_RETRY_LOCKED = "pinRetryLocked"

    const val HEARTBEAT_INTERVAL_MS = 30_000L
    const val FOREGROUND_USAGE_UPLOAD_INTERVAL_MS = 10_000L
    const val FOREGROUND_USAGE_EXTRAPOLATION_MAX_MS = 20_000L
    const val PAIRING_TTL_MS = 10 * 60_000L
    const val TEMP_UNLOCK_MS = 10 * 60_000L
    const val UNLOCK_15_MINUTES_MS = 15 * 60_000L
    const val UNLOCK_30_MINUTES_MS = 30 * 60_000L
    const val SAFE_MODE_DURATION_MS = 30 * 60_000L
    const val TAMPER_EVENT_THROTTLE_MS = 15 * 60_000L
    const val COMMAND_OPEN_SETUP_TTL_MS = 60_000L
    const val COMMAND_STANDARD_TTL_MS = 5 * 60_000L
    const val COMMAND_UNPAIR_TTL_MS = 10 * 60_000L

    fun commandTtlMs(type: String): Long = when (type) {
        COMMAND_OPEN_SETUP -> COMMAND_OPEN_SETUP_TTL_MS
        COMMAND_UNPAIR -> COMMAND_UNPAIR_TTL_MS
        else -> COMMAND_STANDARD_TTL_MS
    }

    val sourceLockPackages = setOf(
        "com.android.tv"
    )

    val sourceLockRuntimePackages = sourceLockPackages + setOf(
        "com.droidlogic.tvinput"
    )

    val primarySettingsPackages = setOf(
        "com.android.tv.settings",
        "com.android.settings"
    )

    val settingsSectionPolicies = listOf(
        SettingsSectionPolicy(
            packageName = SETTINGS_APPS_PACKAGE,
            label = "Settings: Apps",
            shortLabel = "Apps",
            key = "settings-apps"
        ),
        SettingsSectionPolicy(
            packageName = SETTINGS_DEVICE_PREFERENCES_PACKAGE,
            label = "Settings: Device Preferences",
            shortLabel = "Device Preferences",
            key = "settings-device-preferences"
        ),
        SettingsSectionPolicy(
            packageName = SETTINGS_DEVELOPER_OPTIONS_PACKAGE,
            label = "Settings: Developer options",
            shortLabel = "Developer options",
            key = "settings-developer-options"
        ),
        SettingsSectionPolicy(
            packageName = SETTINGS_SECURITY_RESTRICTIONS_PACKAGE,
            label = "Settings: Security & restrictions",
            shortLabel = "Security & restrictions",
            key = "settings-security-restrictions"
        ),
        SettingsSectionPolicy(
            packageName = SETTINGS_ACCESSIBILITY_PACKAGE,
            label = "Settings: Accessibility",
            shortLabel = "Accessibility",
            key = "settings-accessibility"
        ),
        SettingsSectionPolicy(
            packageName = SETTINGS_RESET_PACKAGE,
            label = "Settings: Reset",
            shortLabel = "Reset",
            key = "settings-reset"
        )
    )

    val settingsSectionLockPackages = settingsSectionPolicies.map { it.packageName }.toSet()

    val virtualPolicyPackages = settingsSectionLockPackages

    val deprecatedVirtualPolicyPackages = setOf(
        DEPRECATED_SETTINGS_SECTIONS_PACKAGE
    )

    val parentVisibleLockPackages = sourceLockPackages + primarySettingsPackages + settingsSectionLockPackages

    val defaultLockedPackages = settingsSectionLockPackages

    val alwaysProtectedPackages = setOf(
        TV_PACKAGE,
        "com.android.systemui",
        "com.android.settings",
        "com.android.tv.settings",
        "com.android.packageinstaller",
        "com.google.android.packageinstaller",
        "com.android.documentsui",
        "com.android.permissioncontroller",
        "com.google.android.permissioncontroller",
        "com.google.android.gms",
        "com.google.android.gsf",
        "com.google.android.tvlauncher",
        "com.google.android.apps.tv.launcherx"
    )

    val riskySettingsPackages = emptySet<String>()

    fun isDefaultLocked(packageName: String): Boolean = packageName in defaultLockedPackages

    fun settingsSectionPolicy(packageName: String): SettingsSectionPolicy? {
        return settingsSectionPolicies.firstOrNull { it.packageName == packageName }
    }

    fun settingsSectionPolicyForKey(key: String): SettingsSectionPolicy? {
        return settingsSectionPolicies.firstOrNull { it.key == key }
    }

    fun sourceLockPolicyPackage(packageName: String): String? {
        return if (packageName in sourceLockRuntimePackages) sourceLockPackages.first() else null
    }
}

File: web/js/policyConstants.js
/**
 * Port of shared/PolicyConstants.kt — every constant the parent UI renders or
 * the write paths embed. Keep values string-identical to the Kotlin source.
 */

export const TV_PACKAGE = 'com.guardpulse.parentcontrol.tv';
export const PARENT_PACKAGE = 'com.guardpulse.parentcontrol.parent';

export const SETTINGS_APPS_PACKAGE = 'com.guardpulse.policy.settings_apps';
export const SETTINGS_DEVICE_PREFERENCES_PACKAGE = 'com.guardpulse.policy.settings_device_preferences';
export const SETTINGS_DEVELOPER_OPTIONS_PACKAGE = 'com.guardpulse.policy.settings_developer_options';
export const SETTINGS_SECURITY_RESTRICTIONS_PACKAGE = 'com.guardpulse.policy.settings_security_restrictions';
export const SETTINGS_ACCESSIBILITY_PACKAGE = 'com.guardpulse.policy.settings_accessibility';
export const SETTINGS_RESET_PACKAGE = 'com.guardpulse.policy.settings_reset';
export const DEPRECATED_SETTINGS_SECTIONS_PACKAGE = 'com.guardpulse.policy.settings_sections';

// (packageName, label, shortLabel, unlock key) — order drives parent UI lists.
export const SETTINGS_SECTION_POLICIES = [
  { packageName: SETTINGS_APPS_PACKAGE, label: 'Settings: Apps', shortLabel: 'Apps', key: 'settings-apps' },
  { packageName: SETTINGS_DEVICE_PREFERENCES_PACKAGE, label: 'Settings: Device Preferences', shortLabel: 'Device Preferences', key: 'settings-device-preferences' },
  { packageName: SETTINGS_DEVELOPER_OPTIONS_PACKAGE, label: 'Settings: Developer options', shortLabel: 'Developer options', key: 'settings-developer-options' },
  { packageName: SETTINGS_SECURITY_RESTRICTIONS_PACKAGE, label: 'Settings: Security & restrictions', shortLabel: 'Security & restrictions', key: 'settings-security-restrictions' },
  { packageName: SETTINGS_ACCESSIBILITY_PACKAGE, label: 'Settings: Accessibility', shortLabel: 'Accessibility', key: 'settings-accessibility' },
  { packageName: SETTINGS_RESET_PACKAGE, label: 'Settings: Reset', shortLabel: 'Reset', key: 'settings-reset' },
];

export const COMMAND_RESCAN_APPS = 'rescanApps';
export const COMMAND_RESET_TODAY = 'resetToday';
export const COMMAND_UNPAIR = 'unpair';
export const COMMAND_OPEN_SETUP = 'openSetup';

export const BLOCK_REASON_MANUAL = 'manual';
export const BLOCK_REASON_DAILY_LIMIT = 'dailyLimit';
export const BLOCK_REASON_RISKY_SETTINGS = 'riskySettings';
export const BLOCK_REASON_NETWORK_FILTER_MISSING = 'networkFilterMissing';
export const BLOCK_REASON_SOURCE_LOCK = 'sourceLock';
export const BLOCK_REASON_SETTINGS_SECTION = 'settingsSection';

export const ENFORCEMENT_DEVICE_OWNER = 'deviceOwner';
export const ENFORCEMENT_FALLBACK = 'fallback';
export const ENFORCEMENT_UNPROTECTED = 'unprotected';

export const UNLOCK_PENDING = 'pending';
export const UNLOCK_APPROVED = 'approved';
export const UNLOCK_DENIED = 'denied';
export const UNLOCK_EXPIRED = 'expired';
export const UNLOCK_APPROVAL_ONE_VISIT = 'oneVisit';
export const UNLOCK_APPROVAL_TIMED = 'timed';

export const SYNC_PROTOCOL_VERSION = 2;
export const SYNC_STATUS_APPLIED = 'applied';
export const SYNC_STATUS_FAILED = 'failed';
export const COMMAND_PENDING = 'pending';
export const COMMAND_RUNNING = 'running';
export const COMMAND_DONE = 'done';
export const COMMAND_FAILED = 'failed';
export const COMMAND_EXPIRED = 'expired';

export const PAIR_PENDING = 'pending';
export const PAIR_ACCEPTED = 'accepted';
export const PAIR_REJECTED = 'rejected';
export const PAIR_EXPIRED = 'expired';
export const PAIR_FAILED = 'failed';

export const REVISION_APP_POLICY = 'appPolicy';
export const REVISION_MODE_CREATE = 'modeCreate';
export const REVISION_MODE_UPDATE = 'modeUpdate';
export const REVISION_MODE_DELETE = 'modeDelete';
export const REVISION_MODE_POLICY = 'modePolicy';
export const REVISION_ACTIVE_MODE = 'activeMode';
export const REVISION_SAFE_MODE = 'safeMode';
export const REVISION_PIN = 'pin';
export const REVISION_MIGRATION = 'migration';
export const REVISION_KINDS = new Set([
  REVISION_APP_POLICY, REVISION_MODE_CREATE, REVISION_MODE_UPDATE, REVISION_MODE_DELETE,
  REVISION_MODE_POLICY, REVISION_ACTIVE_MODE, REVISION_SAFE_MODE, REVISION_PIN, REVISION_MIGRATION,
]);

export const TAMPER_ADMIN_DISABLE_REQUESTED = 'adminDisableRequested';
export const TAMPER_ADMIN_DISABLED = 'adminDisabled';
export const TAMPER_ACCESSIBILITY_DISABLED = 'accessibilityDisabled';
export const TAMPER_USAGE_ACCESS_MISSING = 'usageAccessMissing';
export const TAMPER_VPN_DISABLED = 'vpnDisabled';
export const TAMPER_RISKY_SETTINGS_OPENED = 'riskySettingsOpened';
export const TAMPER_PIN_RETRY_LOCKED = 'pinRetryLocked';

export const HEARTBEAT_INTERVAL_MS = 30_000;
export const FOREGROUND_USAGE_UPLOAD_INTERVAL_MS = 10_000;
export const FOREGROUND_USAGE_EXTRAPOLATION_MAX_MS = 20_000;
export const PAIRING_TTL_MS = 600_000;
export const TEMP_UNLOCK_MS = 600_000;
export const UNLOCK_15_MINUTES_MS = 900_000;
export const UNLOCK_30_MINUTES_MS = 1_800_000;
export const SAFE_MODE_DURATION_MS = 1_800_000;
export const TAMPER_EVENT_THROTTLE_MS = 900_000;
export const COMMAND_OPEN_SETUP_TTL_MS = 60_000;
export const COMMAND_STANDARD_TTL_MS = 300_000;
export const COMMAND_UNPAIR_TTL_MS = 600_000;

export function commandTtlMs(type) {
  if (type === COMMAND_OPEN_SETUP) return COMMAND_OPEN_SETUP_TTL_MS;
  if (type === COMMAND_UNPAIR) return COMMAND_UNPAIR_TTL_MS;
  return COMMAND_STANDARD_TTL_MS;
}

export const SOURCE_LOCK_PACKAGES = new Set(['com.android.tv']);
export const SOURCE_LOCK_RUNTIME_PACKAGES = new Set([...SOURCE_LOCK_PACKAGES, 'com.droidlogic.tvinput']);
export const PRIMARY_SETTINGS_PACKAGES = new Set(['com.android.tv.settings', 'com.android.settings']);
export const SETTINGS_SECTION_LOCK_PACKAGES = new Set(SETTINGS_SECTION_POLICIES.map((s) => s.packageName));
export const VIRTUAL_POLICY_PACKAGES = new Set(SETTINGS_SECTION_LOCK_PACKAGES);
export const DEPRECATED_VIRTUAL_POLICY_PACKAGES = new Set([DEPRECATED_SETTINGS_SECTIONS_PACKAGE]);
export const PARENT_VISIBLE_LOCK_PACKAGES = new Set([
  ...SOURCE_LOCK_PACKAGES, ...PRIMARY_SETTINGS_PACKAGES, ...SETTINGS_SECTION_LOCK_PACKAGES,
]);
export const DEFAULT_LOCKED_PACKAGES = new Set(SETTINGS_SECTION_LOCK_PACKAGES);

export const ALWAYS_PROTECTED_PACKAGES = new Set([
  TV_PACKAGE,
  'com.android.systemui',
  'com.android.settings',
  'com.android.tv.settings',
  'com.android.packageinstaller',
  'com.google.android.packageinstaller',
  'com.android.documentsui',
  'com.android.permissioncontroller',
  'com.google.android.permissioncontroller',
  'com.google.android.gms',
  'com.google.android.gsf',
  'com.google.android.tvlauncher',
  'com.google.android.apps.tv.launcherx',
]);

export const RISKY_SETTINGS_PACKAGES = new Set([]);

export function isDefaultLocked(packageName) {
  return DEFAULT_LOCKED_PACKAGES.has(packageName);
}

export function settingsSectionPolicy(packageName) {
  return SETTINGS_SECTION_POLICIES.find((s) => s.packageName === packageName) ?? null;
}

export function settingsSectionPolicyForKey(key) {
  return SETTINGS_SECTION_POLICIES.find((s) => s.key === key) ?? null;
}

export function sourceLockPolicyPackage(packageName) {
  return SOURCE_LOCK_RUNTIME_PACKAGES.has(packageName) ? 'com.android.tv' : null;
}
